kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,472 @@
+      LOG.info("{}[{}] No active segments for initial streamTime {}",
+               name, kafkaPartition, metaDoc.streamTime);
+    } else {
+      for (final SegmentPartition segmentToCreate : activeSegments) {
+        // Most likely if we have active segments upon initialization, the corresponding
+        // collections and windowStart indices already exist, in which case #getCollection and
+        // #createIndex should be quick and won't rebuild either structure if one already exists
+        // TODO: optimize by using listCollectionNames to skip redundant collection/index creation
+        //  since we should almost always have an index for each collection for all active segments
+        partitionSegments.createSegment(segmentToCreate);
+      }
+
+      final long firstSegmentId = activeSegments.get(0).segmentId;
+      LOG.info("{}[{}] Initialized active segments in range {} - {}",
+               name, kafkaPartition, firstSegmentId, firstSegmentId + activeSegments.size());
+    }
+
+    kafkaPartitionToSegments.put(
+        kafkaPartition,
+        partitionSegments
+    );
+
+    return new MongoWindowFlushManager(
+        this,
+        (segment) -> windowsForSegmentPartition(kafkaPartition, segment),
+        metadata,
+        partitioner,
+        kafkaPartition
+    );
+  }
+
+  public String collectionNameForSegment(final SegmentPartition segmentPartition) {
+    return String.format(
+        "%d-%d",
+        segmentPartition.tablePartition, segmentPartition.segmentId
+    );
+  }
+
+  public MongoCollection<WindowDoc> windowsForSegmentPartition(
+      final int kafkaPartition,
+      final SegmentPartition segment
+  ) {
+    return kafkaPartitionToSegments.get(kafkaPartition).segmentWindows.get(segment);
+  }
+
+  public RemoteWriteResult<SegmentPartition> preCommit(
+      final int kafkaPartition
+  ) {
+    final var partitionSegments = kafkaPartitionToSegments.get(kafkaPartition);
+
+    final var segmentBatch = partitionSegments.segmentBatch;
+    final SegmentRoll pendingRoll = partitioner.rolledSegments(
+        name, segmentBatch.flushedStreamTime, segmentBatch.batchStreamTime
+    );
+
+    segmentBatch.prepareRoll(pendingRoll);
+
+    for (final long segmentId : pendingRoll.segmentsToCreate) {
+      partitionSegments.createSegment(new SegmentPartition(kafkaPartition, segmentId));
+    }
+    return RemoteWriteResult.success(null);
+  }
+
+  public RemoteWriteResult<SegmentPartition> postCommit(
+      final int kafkaPartition
+  ) {
+    final var segments = kafkaPartitionToSegments.get(kafkaPartition);
+
+    for (final long segmentId : segments.segmentBatch.segmentRoll.segmentsToExpire) {
+      segments.expireSegment(new SegmentPartition(kafkaPartition, segmentId));
+    }
+    segments.segmentBatch.finalizeRoll();
+    return RemoteWriteResult.success(null);
+  }
+
+  @Override
+  public long fetchOffset(final int kafkaPartition) {
+    final WindowMetadataDoc result = metadata.find(
```

Comment:
The `fetchOffset` (maybe better named "fetchStartOffset" or "fetchRestoreOffset") is always invoked just a few lines after the `init` call, so I think it would be rare to get fenced between the `init` and `fetchOffset` calls. Being fenced by offset would be next to impossible since it would require another client to start up, overwrite the epoch, process a batch of records and then flush them to bump the offset, all in the time between when the table is initialized and when we read out the starting offset to restore from. 

An epoch mismatch is probably rare as well, but that case is at least possible. I suppose it doesn't hurt to check the epoch here. I checked the Streams callstack and it looks like we can safely throw TaskMigratedException from here and have it interpreted correctly so that's good